Move the JupyterLab registry token to the base package.

jasongrout opened the pull request:
Addresses #1781 

I looked into moving the token to the base package. I realized a few things:

1. There is a lot of confusion right now in what the jlab plugin exports. It exports the actual jlab document widget, but it shouldn't - it should just export the registration function. This should definitely be untangled.
2. The registry is specific to the jlab widget manager - the classic notebook widget manager does not have the concept of a registry (it uses require as a default registry). So it's a little weird to have the concept of a registry in the base package and/or the base widget manager.

This is definitely still a work in progress.

blink1073 commented:
We could.define the interface and token I base.for a jlab manager, and have the implementation in a separate package. That is what we do for the statedb for example.

blink1073 commented:
Actually the default statedb class is in the same file but the point stands. 😁

jasongrout commented:
> We could.define the interface and token in base for a jlab manager, and have the implementation in a separate package.

Yes, that's essentially what the current PR state does. But that's what I meant by it being weird to have the concept (i.e., interface) for a registry in the base package, because the most prevalent widget manager does not have the concept of a registry (it punts to require to do that).

blink1073 commented:
We could call it a JupyterLab registry.  It doesn't tie the package to JupyterLab, but makes it explicit what it is intended to be used for.

jasongrout commented:
Yes. It still feels weird to have jlab specific stuff in the base package, but I think we'll probably do it as a matter of practicality for widget library authors.

blink1073 commented:
Doesn't seem weird enough to have a completely separate package, given that it doesn't (in effect) add any dependencies and to the base package and is only an interface and a runtime representation of an interface.

blink1073 commented:
It could be made more generic:

```typescript
/**
 * A runtime interface object for a widget registry.
 */
export
const IJupyterWidgetRegistry = new Token<IJupyterWidgetRegistry>('jupyter.extensions.jupyterWidgetRegistry');


/**
 * A registry of Jupyter Widgets.  Used by widget managers that support
 * an external registry.
 */
export
interface IJupyterWidgetRegistry {
  /**
   * Register a widget module.
   */
  registerWidget(data: WidgetManager.IWidgetData): void;
}
```

jasongrout commented:
> It could be made more generic:

That's almost exactly what is in this PR: https://github.com/jupyter-widgets/ipywidgets/pull/1789/files#diff-121363ec153b2a74f364564bd76b8fd5

I like how your docs indicate "for those widget managers that use a registry".

blink1073 commented:
Yeah I cribbed it from the PR 😉

jasongrout commented:
@blink1073 - I rebased and squashed the changes. Let me know if it looks good to you. If it does, I'll merge and release a new version of the jlab manager.

jasongrout commented:
Marking as 7.0.x since it just changes the experimental jlab interface.

blink1073 commented:
Looks great, thanks!

jasongrout commented:
tests were failing - your dependency-checking script caught the no-longer-necessary dependency. Thanks again for that!

blink1073 commented:
👊

jasongrout commented:
Just released `@jupyter-widgets/jupyterlab-manager` 0.29 and `@jupyter-widgets/base` 1.1 with this change.
